src/aiida/orm/nodes/data/array/kpoints.py

```diff
@@ -223,16 +223,26 @@ def set_cell_from_structure(self, structuredata):
 
         :param structuredata: an instance of StructureData
         """
-        from aiida.orm import StructureData
+        from aiida.orm.nodes.data.structure import StructureData, has_atomistic
 
         if not isinstance(structuredata, StructureData):
-            raise ValueError(
-                'An instance of StructureData should be passed to ' 'the KpointsData, found instead {}'.format(
-                    structuredata.__class__
-                )
+            error_message = (
+                'An instance of StructureData or aiida-atomistic StructureData should be passed to '
+                'the KpointsData, found instead {}'.format(structuredata.__class__)
             )
-        cell = structuredata.cell
-        self.set_cell(cell, structuredata.pbc)
+            if has_atomistic():
+                from aiida_atomistic import StructureData as AtomisticStructureData
+
+                if not isinstance(structuredata, AtomisticStructureData):
+                    raise ValueError(error_message)
+                else:
+                    cell = structuredata.cell
+                    self.set_cell(cell, structuredata.pbc)
+            else:
+                raise ValueError(error_message)
```

Comment:
Maybe it is possible to move the exception to same condition and move the `set_cell` to same condition.

Reply:
Hi @unkcpz, thank you for the comment. I have actually another idea, which is replacing the whole block from line 228 to 242 with:
```python
if has_atomistic():
    from aiida_atomistic import StructureData as AtomisticStructureData

if not isinstance(structuredata, StructureData) and not isinstance(structuredata, AtomisticStructureData if has_atomistic() else StructureData):
    raise ValueError('An instance of StructureData or aiida-atomistic StructureData should be passed to '
                'the KpointsData, found instead {}'.format(structuredata.__class__))
else:
    cell = structuredata.cell
    self.set_cell(cell, structuredata.pbc)
```

This seems redundant, but actually makes more compact the check. What do you think?

Reply:
I guess if the `structuredata` is neither `StructureData` nor the `AtomisticStructureData`, the else block of calling `structuredata.cell` and `self.set_cell` with `structuredata.pbc` will fail? 
If so, why not use try..except block and put the current else block in the try and the catch the exceptions for correspond operations? Let me know if I am missing something and we can try it together in person.

Reply:
In the way you suggest, if `structuredata` was an arbitrary object which defines `cell` and `pbc` as attributes, the exceptions will not be raised but it should. And the user will not see the error (i.e. not providing a an orm or atomistic `StructureData`).

Anyway, let's discuss this in person on monday

Reply:
@mikibonacci mentioned in aiida-pseudo there is a similar suggestion by Seb that will work here.